File: app/src/main/java/name/monwf/customiuizer/mods/utils/ModuleHelper.java
package name.monwf.customiuizer.mods.utils;

import static name.monwf.customiuizer.mods.GlobalActions.ACTION_PREFIX;
import static name.monwf.customiuizer.mods.utils.XposedHelpers.findClass;
import static name.monwf.customiuizer.mods.utils.XposedHelpers.log;

import android.annotation.SuppressLint;
import android.app.ActivityOptions;
import android.app.Application;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.content.res.Configuration;
import android.content.res.Resources;
import android.net.Uri;
import android.os.Bundle;
import android.os.Process;
import android.os.UserHandle;
import android.provider.Settings;
import android.util.MiuiMultiWindowUtils;
import android.view.View;

import androidx.annotation.Nullable;

import java.io.RandomAccessFile;
import java.lang.reflect.Method;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;

import io.github.libxposed.api.XposedModuleInterface;
import miui.app.MiuiFreeFormManager;
import miui.process.ForegroundInfo;
import miui.process.ProcessManager;
import name.monwf.customiuizer.MainModule;
import name.monwf.customiuizer.mods.utils.HookerClassHelper.CustomMethodUnhooker;
import name.monwf.customiuizer.mods.utils.HookerClassHelper.MethodHook;
import name.monwf.customiuizer.utils.Helpers;


public class ModuleHelper {
    public static final String NOT_EXIST_SYMBOL = "ObjectFieldNotExist";

    public static final String prefsName = "customiuizer_prefs";

    public static String currentPackageName;

    @SuppressLint("StaticFieldLeak")
    private static Context mModuleContext = null;

    private final static int viewInfoTag = ResourceHooks.getFakeResId("view_info_tag");

    static HashSet<PreferenceObserver> prefObservers = new HashSet<PreferenceObserver>();

    static Class<?> ActivityThreadClass;

    static {
        ActivityThreadClass = null;
    }

    public static void printCallStack() {
        StackTraceElement[] stackTrace = Thread.currentThread().getStackTrace();
        for (StackTraceElement el: stackTrace)
            if (el != null) {
                log(el.getClassName() + " $$ " + el.getMethodName());
            }
    }

    public static CustomMethodUnhooker hookMethod(Method method, MethodHook callback) {
        try {
            return XposedHelpers.doHookMethod(method, callback);
        } catch (Throwable t) {
            log("Failed to hook " + method.getName() + " method");
            return null;
        }
    }

    public static CustomMethodUnhooker findAndHookMethod(String className, ClassLoader classLoader, String methodName, Object... parameterTypesAndCallback) {
        try {
            return XposedHelpers.findAndHookMethod(className, classLoader, methodName, parameterTypesAndCallback);
        } catch (Throwable t) {
            log("Failed to hook " + methodName + " method in " + className);
            return null;
        }
    }

    /**
     * Calls an instance or static method of the given object silently.
     *
     * @param obj        The object instance. A class reference is not sufficient!
     * @param methodName The method name.
     * @param args       The arguments for the method call.
     */
    public static Object callMethodSilently(Object obj, String methodName, Object... args) {
        try {
            return XposedHelpers.callMethod(obj, methodName, args);
        } catch (Throwable e) {
            XposedHelpers.log(e);
            return NOT_EXIST_SYMBOL;
        }
    }

    public static CustomMethodUnhooker findAndHookMethod(Class<?> clazz, String methodName, Object... parameterTypesAndCallback) {
        try {
            return XposedHelpers.findAndHookMethod(clazz, methodName, parameterTypesAndCallback);
        } catch (Throwable t) {
            log("Failed to hook " + methodName + " method in " + clazz.getCanonicalName());
            return null;
        }
    }

    @SuppressWarnings("UnusedReturnValue")
    public static boolean findAndHookMethodSilently(String className, ClassLoader classLoader, String methodName, Object... parameterTypesAndCallback) {
        try {
            XposedHelpers.findAndHookMethod(className, classLoader, methodName, parameterTypesAndCallback);
            return true;
        } catch (Throwable t) {
            return false;
        }
    }

    @SuppressWarnings("UnusedReturnValue")
    public static boolean findAndHookMethodSilently(Class<?> clazz, String methodName, Object... parameterTypesAndCallback) {
        try {
            XposedHelpers.findAndHookMethod(clazz, methodName, parameterTypesAndCallback);
            return true;
        } catch (Throwable t) {
            return false;
        }
    }

    public static CustomMethodUnhooker findAndHookConstructor(String className, ClassLoader classLoader, Object... parameterTypesAndCallback) {
        try {
            return XposedHelpers.findAndHookConstructor(className, classLoader, parameterTypesAndCallback);
        } catch (Throwable t) {
            log("Failed to hook constructor in " + className);
            return null;
        }
    }

    public static void hookAllConstructors(String className, ClassLoader classLoader, MethodHook callback) {
        try {
            Class<?> hookClass = XposedHelpers.findClassIfExists(className, classLoader);
            if (hookClass == null || XposedHelpers.hookAllConstructors(hookClass, callback).size() == 0)
                log("Failed to hook " + className + " constructor");
        } catch (Throwable t) {
            log(t);
        }
    }

    public static void hookAllConstructors(Class<?> hookClass, MethodHook callback) {
        try {
            if (XposedHelpers.hookAllConstructors(hookClass, callback).size() == 0)
                log("Failed to hook " + hookClass.getCanonicalName() + " constructor");
        } catch (Throwable t) {
            log(t);
        }
    }

    public static void hookAllMethods(String className, ClassLoader classLoader, String methodName, MethodHook callback) {
        try {
            Class<?> hookClass = XposedHelpers.findClassIfExists(className, classLoader);
            if (hookClass == null || XposedHelpers.hookAllMethods(hookClass, methodName, callback).size() == 0)
                log("Failed to hook " + methodName + " method in " + className);
        } catch (Throwable t) {
            log(t);
        }
    }

    public static void hookAllMethods(Class<?> hookClass, String methodName, MethodHook callback) {
        try {
            if (XposedHelpers.hookAllMethods(hookClass, methodName, callback).size() == 0)
                log("Failed to hook " + methodName + " method in " + hookClass.getCanonicalName());
        } catch (Throwable t) {
            log(t);
        }
    }

    public static Object proxySystemProperties(String method, String prop, String val, ClassLoader classLoader) {
        return XposedHelpers.callStaticMethod(XposedHelpers.findClassIfExists("android.os.SystemProperties", classLoader),
            method, prop, val);
    }

    public static Object proxySystemProperties(String method, String prop, int val, ClassLoader classLoader) {
        return XposedHelpers.callStaticMethod(XposedHelpers.findClassIfExists("android.os.SystemProperties", classLoader),
            method, prop, val);
    }

    public static boolean hookAllMethodsSilently(String className, ClassLoader classLoader, String methodName, MethodHook callback) {
        try {
            Class<?> hookClass = XposedHelpers.findClassIfExists(className, classLoader);
            return hookClass != null && XposedHelpers.hookAllMethods(hookClass, methodName, callback).size() > 0;
        } catch (Throwable t) {
            return false;
        }
    }

    public static boolean hookAllMethodsSilently(Class<?> hookClass, String methodName, MethodHook callback) {
        try {
            return hookClass != null && XposedHelpers.hookAllMethods(hookClass, methodName, callback).size() > 0;
        } catch (Throwable t) {
            return false;
        }
    }

    public static Object getStaticObjectFieldSilently(Class <?> clazz, String fieldName) {
        try {
            return XposedHelpers.getStaticObjectField(clazz, fieldName);
        } catch (Throwable t) {
            return NOT_EXIST_SYMBOL;
        }
    }

    public static Object getObjectFieldSilently(Object obj, String fieldName) {
        try {
            return XposedHelpers.getObjectField(obj, fieldName);
        } catch (Throwable t) {
            return NOT_EXIST_SYMBOL;
        }
    }

    public static int getUserId() {
        return (int)XposedHelpers.callStaticMethod(UserHandle.class, "getUserId", Process.myUid());
    }


    public static Context findContext() {
        Context context = null;
        try {
            if (ActivityThreadClass == null) {
                ActivityThreadClass = XposedHelpers.findClass("android.app.ActivityThread", null);
            }
            context = (Application)XposedHelpers.callStaticMethod(ActivityThreadClass, "currentApplication");
            if (context == null) {
                Object currentActivityThread = XposedHelpers.callStaticMethod(ActivityThreadClass, "currentActivityThread");
                if (currentActivityThread != null) context = (Context)XposedHelpers.callMethod(currentActivityThread, "getSystemContext");
            }
        } catch (Throwable ignore) {}
        return context;
    }

    public static Context findContext(XposedModuleInterface.PackageLoadedParam lpparam) {
        Context context = null;
        try {
            context = (Application)XposedHelpers.callStaticMethod(XposedHelpers.findClass("android.app.ActivityThread", lpparam.getClassLoader()), "currentApplication");
            if (context == null) {
                Object currentActivityThread = XposedHelpers.callStaticMethod(XposedHelpers.findClass("android.app.ActivityThread", null), "currentActivityThread");
                if (currentActivityThread != null) context = (Context)XposedHelpers.callMethod(currentActivityThread, "getSystemContext");
            }
        } catch (Throwable ignore) {}
        return context;
    }

    public static String stringifyBundle(Bundle bundle) {
        if (bundle == null) {
            return null;
        }
        StringBuilder string = new StringBuilder("Bundle{");
        for (String key : bundle.keySet()) {
            string.append(" ").append(key).append(" -> ").append(bundle.get(key)).append(";");
        }
        string.append(" }Bundle");
        return string.toString();
    }

    public static long getNextMIUIAlarmTime(Context context) {
        long nextTime = 0;
        try {
            nextTime = Settings.Global.getLong(context.getContentResolver(), "next_alarm_clock_long");
        } catch (Settings.SettingNotFoundException e) {}
        return nextTime;
    }
    public static void openAppInfo(Context context, String pkg, int user) {
        try {
            Intent intent = new Intent("miui.intent.action.APP_MANAGER_APPLICATION_DETAIL");
            intent.setPackage("com.miui.securitycenter");
            intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_RESET_TASK_IF_NEEDED);
            intent.putExtra("package_name", pkg);
            if (user != 0) intent.putExtra("miui.intent.extra.USER_ID", user);
            context.startActivity(intent);
        } catch (Throwable t) {
            try {
                Intent intent = new Intent(android.provider.Settings.ACTION_APPLICATION_DETAILS_SETTINGS);
                intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_RESET_TASK_IF_NEEDED);
                intent.setData(Uri.parse("package:" + pkg));
                if (user != 0)
                    XposedHelpers.callMethod(context, "startActivityAsUser", intent, XposedHelpers.newInstance(UserHandle.class, user));
                else
                    context.startActivity(intent);
            } catch (Throwable t2) {
                log(t2);
            }
        }
    }

    public interface PreferenceObserver {
        void onChange(String key);
    }

    public static void observePreferenceChange(PreferenceObserver prefObserver) {
        prefObservers.add(prefObserver);
    }

    public static void handlePreferenceChanged(@Nullable String key) {
        for (PreferenceObserver prefObserver:prefObservers) {
            prefObserver.onChange(key);
        }
    }

    public static synchronized Context getModuleContext(Context context) throws Throwable {
        return getModuleContext(context, null);
    }

    public static synchronized Context getModuleContext(Context context, Configuration config) throws Throwable {
        if (mModuleContext == null) {
            mModuleContext = context.createPackageContext(Helpers.modulePkg, Context.CONTEXT_IGNORE_SECURITY);
        }
        return config == null ? mModuleContext : mModuleContext.createConfigurationContext(config);
    }

    public static synchronized Resources getModuleRes(Context context) throws Throwable {
        Configuration config = context.getResources().getConfiguration();
        Context moduleContext = getModuleContext(context, config);
        return moduleContext.getResources();
    }

    public static Object getDepInstance(ClassLoader classLoader, String className) {
        Class<?> DependencyClass = findClass("com.android.systemui.Dependency", classLoader);
        return XposedHelpers.callStaticMethod(DependencyClass, "get", findClass(className, classLoader));
    }

    public static Object getViewInfo(View view, String key) {
        Object info = view.getTag(viewInfoTag);
        if (info == null) {
            return null;
        }
        HashMap<String, Object> viewInfo;
        viewInfo = (HashMap<String, Object>) info;
        return viewInfo.get(key);
    }
    public static void setViewInfo(View view, String key, Object value) {
        Object info = view.getTag(viewInfoTag);
        HashMap<String, Object> viewInfo;
        if (info == null) {
            viewInfo = new HashMap<String, Object>();
            view.setTag(viewInfoTag, viewInfo);
        }
        else {
            viewInfo = (HashMap<String, Object>) info;
        }
        viewInfo.put(key, value);
    }
    public static Bundle getFreeformOptions(Context mContext, String pkgName, PendingIntent pendingIntent, boolean ignoreCheck) throws PendingIntent.CanceledException {
        if (!ignoreCheck) {
            ForegroundInfo foregroundInfo = ProcessManager.getForegroundInfo();
            if (foregroundInfo != null) {
                String topPackage = foregroundInfo.mForegroundPackageName;
                if (pkgName.equals(topPackage)) {
                    return null;
                }
            }
            List<MiuiFreeFormManager.MiuiFreeFormStackInfo> freeFormStackInfoList = MiuiFreeFormManager.getAllFreeFormStackInfosOnDisplay(mContext.getDisplay() != null ? mContext.getDisplay().getDisplayId() : 0);
            int freeFormCount = 0;
            if (freeFormStackInfoList != null) {
                freeFormCount = freeFormStackInfoList.size();
            }
            if (freeFormCount == 2) return null;
            for (MiuiFreeFormManager.MiuiFreeFormStackInfo rootTaskInfo : freeFormStackInfoList) {
                if (pkgName.equals(rootTaskInfo.packageName)) return null;
            }
        }
        if (!pendingIntent.isActivity()) {
            Intent bIntent = new Intent(ACTION_PREFIX + "SetFreeFormPackage");
            bIntent.putExtra("package", pkgName);
            bIntent.setPackage("android");
            mContext.sendBroadcast(bIntent);
        }
        ActivityOptions options = MiuiMultiWindowUtils.getActivityOptions(mContext, pkgName, true, false);
        if (options != null) {
            XposedHelpers.callMethod(options, "setFreeformAnimation", false);
        }
        return options != null ? options.toBundle() : null;
    }
    public static Intent getFreeformIntent(String pkgName) {
        Intent intent = new Intent();
        if (!"com.tencent.tim".equals(pkgName)) {
            XposedHelpers.callMethod(intent, "addFlags", 134217728);
            XposedHelpers.callMethod(intent, "addFlags", 268435456);
            XposedHelpers.callMethod(intent, "addMiuiFlags", 256);
        }
        return intent;
    }
    private static int thermalId = -1;
    public static int getCPUThermalId() {
        if (thermalId != -1) return thermalId;
        for (var i = 2;i < 40;i = i + 2) {
            try {
                RandomAccessFile cpuReader = new RandomAccessFile("/sys/devices/virtual/thermal/thermal_zone" + i + "/type", "r");
                String sensorType = cpuReader.readLine();
                cpuReader.close();
                if (sensorType.startsWith("cpu-") || sensorType.startsWith("cpu_big")) {
                    thermalId = i;
                    break;
                }
            } catch (Throwable ign) {}
        }
        return thermalId;
    }

    public static void replacePkgAndFrameworkValue(String pkg, String type, String name, Object resValue) {
        if (!"android".equals(pkg)) {
            MainModule.resHooks.setThemeValueReplacement("android", type, name, resValue);
        }
        MainModule.resHooks.setThemeValueReplacement(pkg, type, name, resValue);
    }
    public static Object getObjectFieldByPath(Object target, String path) {
        if (target == null) return null;
        String[] pathes = path.split("\\.");
        for (String field:pathes) {
            target = getObjectFieldSilently(target, field);
            if (NOT_EXIST_SYMBOL.equals(target)) {
                return NOT_EXIST_SYMBOL;
            }
        }
        return target;
    }
}

File: app/src/main/java/tv/withaibuild/customiuizer/mods/utils/ModuleHelper.kt
package tv.withaibuild.customiuizer.mods.utils

import android.annotation.SuppressLint
import android.app.ActivityOptions
import android.app.Application
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.content.res.Configuration
import android.content.res.Resources
import android.net.Uri
import android.os.Bundle
import android.os.Process
import android.os.UserHandle
import android.provider.Settings
import android.util.MiuiMultiWindowUtils
import android.view.View
import io.github.libxposed.api.XposedModuleInterface
import miui.app.MiuiFreeFormManager
import miui.process.ForegroundInfo
import miui.process.ProcessManager
import tv.withaibuild.customiuizer.MainModule
import tv.withaibuild.customiuizer.mods.utils.HookerClassHelper.CustomMethodUnhooker
import tv.withaibuild.customiuizer.mods.utils.HookerClassHelper.MethodHook
import tv.withaibuild.customiuizer.utils.Helpers
import java.io.RandomAccessFile
import java.lang.reflect.Method
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.CopyOnWriteArraySet

class ModuleHelper private constructor() {

    interface PreferenceObserver {
        fun onChange(key: String?)
    }

    companion object {
        const val NOT_EXIST_SYMBOL = "ObjectFieldNotExist"
        const val prefsName = "customiuizer_prefs"

        @JvmField
        var currentPackageName: String? = null

        @SuppressLint("StaticFieldLeak")
        @JvmField
        var mModuleContext: Context? = null

        @SuppressLint("StaticFieldLeak")
        @JvmField
        var mCachedContext: Context? = null

        @JvmField
        var cachedModuleRes: Resources? = null

        @JvmField
        var cachedModuleConfig: Configuration? = null

        private val viewInfoTag = ResourceHooks.getFakeResId("view_info_tag")
        private val prefObservers = CopyOnWriteArraySet<PreferenceObserver>()
        private const val PREF_OBSERVER_FIELD = "customiuizer_prefObserver"

        @JvmField
        internal var ActivityThreadClass: Class<*>? = null

        private val depInstanceCache = ConcurrentHashMap<Class<*>, Any?>()

        private var DependencyClass: Class<*>? = null

        private var DependencyGetMethod: Method? = null

        private var thermalId = -1

        @JvmStatic
        fun printCallStack() {
            val stackTrace = Thread.currentThread().stackTrace
            for (el in stackTrace) {
                if (el != null) {
                    XposedHelpers.log(el.className + " $$ " + el.methodName)
                }
            }
        }

        @JvmStatic
        fun hookMethod(method: Method, callback: MethodHook): CustomMethodUnhooker? {
            return try {
                XposedHelpers.doHookMethod(method, callback)
            } catch (t: Throwable) {
                XposedHelpers.log("Failed to hook " + method.name + " method")
                null
            }
        }

        @JvmStatic
        fun findAndHookMethod(className: String, classLoader: ClassLoader?, methodName: String, vararg parameterTypesAndCallback: Any?): CustomMethodUnhooker? {
            return try {
                XposedHelpers.findAndHookMethod(className, classLoader, methodName, *parameterTypesAndCallback)
            } catch (t: Throwable) {
                XposedHelpers.log("Failed to hook " + methodName + " method in " + className)
                null
            }
        }

        @JvmStatic
        fun callMethodSilently(obj: Any?, methodName: String, vararg args: Any?): Any? {
            return try {
                XposedHelpers.callMethod(obj, methodName, *args)
            } catch (e: Throwable) {
                XposedHelpers.log(e)
                NOT_EXIST_SYMBOL
            }
        }

        @JvmStatic
        fun findAndHookMethod(clazz: Class<*>, methodName: String, vararg parameterTypesAndCallback: Any?): CustomMethodUnhooker? {
            return try {
                XposedHelpers.findAndHookMethod(clazz, methodName, *parameterTypesAndCallback)
            } catch (t: Throwable) {
                XposedHelpers.log("Failed to hook " + methodName + " method in " + clazz?.canonicalName)
                null
            }
        }

        @JvmStatic
        @Suppress("UNUSED_RETURN_VALUE")
        fun findAndHookMethodSilently(className: String, classLoader: ClassLoader?, methodName: String, vararg parameterTypesAndCallback: Any?): Boolean {
            return try {
                XposedHelpers.findAndHookMethod(className, classLoader, methodName, *parameterTypesAndCallback)
                true
            } catch (t: Throwable) {
                false
            }
        }

        @JvmStatic
        @Suppress("UNUSED_RETURN_VALUE")
        fun findAndHookMethodSilently(clazz: Class<*>, methodName: String, vararg parameterTypesAndCallback: Any?): Boolean {
            return try {
                XposedHelpers.findAndHookMethod(clazz, methodName, *parameterTypesAndCallback)
                true
            } catch (t: Throwable) {
                false
            }
        }

        @JvmStatic
        fun findAndHookConstructor(className: String, classLoader: ClassLoader?, vararg parameterTypesAndCallback: Any?): CustomMethodUnhooker? {
            return try {
                XposedHelpers.findAndHookConstructor(className, classLoader, *parameterTypesAndCallback)
            } catch (t: Throwable) {
                XposedHelpers.log("Failed to hook constructor in " + className)
                null
            }
        }

        @JvmStatic
        fun hookAllConstructors(className: String, classLoader: ClassLoader?, callback: MethodHook) {
            try {
                val hookClass = XposedHelpers.findClassIfExists(className, classLoader)
                if (hookClass == null || XposedHelpers.hookAllConstructors(hookClass, callback).isEmpty()) {
                    XposedHelpers.log("Failed to hook " + className + " constructor")
                }
            } catch (t: Throwable) {
                XposedHelpers.log(t)
            }
        }

        @JvmStatic
        fun hookAllConstructors(hookClass: Class<*>?, callback: MethodHook) {
            try {
                if (XposedHelpers.hookAllConstructors(hookClass, callback).isEmpty()) {
                    XposedHelpers.log("Failed to hook " + hookClass?.canonicalName + " constructor")
                }
            } catch (t: Throwable) {
                XposedHelpers.log(t)
            }
        }

        @JvmStatic
        fun hookAllMethods(className: String, classLoader: ClassLoader?, methodName: String, callback: MethodHook) {
            try {
                val hookClass = XposedHelpers.findClassIfExists(className, classLoader)
                if (hookClass == null || XposedHelpers.hookAllMethods(hookClass, methodName, callback).isEmpty()) {
                    XposedHelpers.log("Failed to hook " + methodName + " method in " + className)
                }
            } catch (t: Throwable) {
                XposedHelpers.log(t)
            }
        }

        @JvmStatic
        fun hookAllMethods(hookClass: Class<*>?, methodName: String, callback: MethodHook) {
            try {
                if (XposedHelpers.hookAllMethods(hookClass, methodName, callback).isEmpty()) {
                    XposedHelpers.log("Failed to hook " + methodName + " method in " + hookClass?.canonicalName)
                }
            } catch (t: Throwable) {
                XposedHelpers.log(t)
            }
        }

        @JvmStatic
        fun proxySystemProperties(method: String, prop: String, value: String, classLoader: ClassLoader?): Any? {
            val sysPropClass = XposedHelpers.findClassIfExists("android.os.SystemProperties", classLoader) ?: return null
            return XposedHelpers.callStaticMethod(sysPropClass, method, prop, value)
        }

        @JvmStatic
        fun proxySystemProperties(method: String, prop: String, value: Int, classLoader: ClassLoader?): Any? {
            val sysPropClass = XposedHelpers.findClassIfExists("android.os.SystemProperties", classLoader) ?: return null
            return XposedHelpers.callStaticMethod(sysPropClass, method, prop, value)
        }

        @JvmStatic
        fun hookAllMethodsSilently(className: String, classLoader: ClassLoader?, methodName: String, callback: MethodHook): Boolean {
            return try {
                val hookClass = XposedHelpers.findClassIfExists(className, classLoader)
                hookClass != null && XposedHelpers.hookAllMethods(hookClass, methodName, callback).isNotEmpty()
            } catch (t: Throwable) {
                false
            }
        }

        @JvmStatic
        fun hookAllMethodsSilently(hookClass: Class<*>?, methodName: String, callback: MethodHook): Boolean {
            return try {
                hookClass != null && XposedHelpers.hookAllMethods(hookClass, methodName, callback).isNotEmpty()
            } catch (t: Throwable) {
                false
            }
        }

        @JvmStatic
        fun getStaticObjectFieldSilently(clazz: Class<*>, fieldName: String): Any? {
            return try {
                XposedHelpers.getStaticObjectField(clazz, fieldName)
            } catch (t: Throwable) {
                NOT_EXIST_SYMBOL
            }
        }

        @JvmStatic
        fun getObjectFieldSilently(obj: Any?, fieldName: String): Any? {
            return try {
                XposedHelpers.getObjectField(obj, fieldName)
            } catch (t: Throwable) {
                NOT_EXIST_SYMBOL
            }
        }

        @JvmStatic
        fun getUserId(): Int {
            return Process.myUid() / 100000
        }

        @JvmStatic
        fun findContext(): Context? {
            if (mCachedContext != null) return mCachedContext
            var context: Context? = null
            try {
                val atClass = ActivityThreadClass
                    ?: XposedHelpers.findClass("android.app.ActivityThread", null).also { ActivityThreadClass = it }
                context = XposedHelpers.callStaticMethod(atClass, "currentApplication") as? Application
                if (context == null) {
                    val currentActivityThread = XposedHelpers.callStaticMethod(atClass, "currentActivityThread")
                    if (currentActivityThread != null) {
                        context = XposedHelpers.callMethod(currentActivityThread, "getSystemContext") as? Context
                    }
                }
            } catch (ignore: Throwable) {
            }
            if (context != null) mCachedContext = context
            return context
        }

        @JvmStatic
        fun findContext(lpparam: XposedModuleInterface.PackageReadyParam?): Context? {
            var context: Context? = null
            try {
                val classLoader = lpparam?.classLoader
                context = XposedHelpers.callStaticMethod(
                    XposedHelpers.findClass("android.app.ActivityThread", classLoader),
                    "currentApplication"
                ) as? Application
                if (context == null) {
                    val currentActivityThread = XposedHelpers.callStaticMethod(
                        XposedHelpers.findClass("android.app.ActivityThread", null),
                        "currentActivityThread"
                    )
                    if (currentActivityThread != null) {
                        context = XposedHelpers.callMethod(currentActivityThread, "getSystemContext") as? Context
                    }
                }
            } catch (ignore: Throwable) {
            }
            return context
        }

        @JvmStatic
        fun stringifyBundle(bundle: Bundle?): String? {
            if (bundle == null) return null
            val string = StringBuilder("Bundle{")
            for (key in bundle.keySet()) {
                string.append(" ").append(key).append(" -> ").append(bundle.get(key)).append(";")
            }
            string.append(" }Bundle")
            return string.toString()
        }

        @JvmStatic
        fun getNextMIUIAlarmTime(context: Context): Long {
            var nextTime = 0L
            try {
                nextTime = Settings.Global.getLong(context.contentResolver, "next_alarm_clock_long")
            } catch (e: Settings.SettingNotFoundException) {
            }
            return nextTime
        }

        @JvmStatic
        fun openAppInfo(context: Context, pkg: String, user: Int) {
            try {
                val intent = Intent("miui.intent.action.APP_MANAGER_APPLICATION_DETAIL")
                intent.setPackage("com.miui.securitycenter")
                intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_RESET_TASK_IF_NEEDED
                intent.putExtra("package_name", pkg)
                if (user != 0) intent.putExtra("miui.intent.extra.USER_ID", user)
                context.startActivity(intent)
            } catch (t: Throwable) {
                try {
                    val intent = Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS)
                    intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_RESET_TASK_IF_NEEDED
                    intent.data = Uri.parse("package:$pkg")
                    if (user != 0) {
                        val userHandle = XposedHelpers.newInstance(UserHandle::class.java, user) as UserHandle
                        XposedHelpers.callMethod(context, "startActivityAsUser", intent, userHandle)
                    } else {
                        context.startActivity(intent)
                    }
                } catch (t2: Throwable) {
                    XposedHelpers.log(t2)
                }
            }
        }

        @JvmStatic
        fun observePreferenceChange(prefObserver: PreferenceObserver?) {
            if (prefObserver != null) prefObservers.add(prefObserver)
        }

        @JvmStatic
        fun observePreferenceChange(prefObserver: PreferenceObserver?, owner: Any?) {
            if (prefObserver == null) return
            if (owner == null) {
                observePreferenceChange(prefObserver)
                return
            }
            val old = XposedHelpers.getAdditionalInstanceField(owner, PREF_OBSERVER_FIELD)
            if (old is PreferenceObserver) {
                prefObservers.remove(old)
            }
            XposedHelpers.setAdditionalInstanceField(owner, PREF_OBSERVER_FIELD, prefObserver)
            prefObservers.add(prefObserver)
        }

        @JvmStatic
        fun removePreferenceObserver(owner: Any?) {
            val old = XposedHelpers.removeAdditionalInstanceField(owner, PREF_OBSERVER_FIELD)
            if (old is PreferenceObserver) {
                prefObservers.remove(old)
            }
        }

        @JvmStatic
        fun handlePreferenceChanged(key: String?) {
            for (prefObserver in prefObservers) {
                prefObserver.onChange(key)
            }
        }

        @JvmStatic
        @Synchronized
        @JvmOverloads
        fun getModuleContext(context: Context, config: Configuration? = null): Context {
            if (mModuleContext == null) {
                mModuleContext = context.createPackageContext(Helpers.modulePkg, Context.CONTEXT_IGNORE_SECURITY)
            }
            return if (config == null) mModuleContext!! else mModuleContext!!.createConfigurationContext(config)
        }

        @JvmStatic
        @Synchronized
        fun getModuleRes(context: Context): Resources {
            val newConfig = context.resources.configuration
            if (cachedModuleRes != null && cachedModuleConfig == newConfig) {
                return cachedModuleRes!!
            }
            val config = Configuration(newConfig)
            val moduleContext = getModuleContext(context, config)
            cachedModuleRes = moduleContext.resources
            cachedModuleConfig = config
            return cachedModuleRes!!
        }

        @JvmStatic
        fun getDepInstance(classLoader: ClassLoader?, className: String): Any? {
            return try {
                val clazz = XposedHelpers.findClass(className, classLoader)
                val cached = depInstanceCache[clazz]
                if (cached != null) return cached

                if (DependencyClass == null || DependencyClass?.classLoader != classLoader) {
                    DependencyClass = XposedHelpers.findClass("com.android.systemui.Dependency", classLoader)
                    DependencyGetMethod = DependencyClass?.getDeclaredMethod("get", Class::class.java)
                    DependencyGetMethod?.isAccessible = true
                }
                val instance = DependencyGetMethod?.invoke(null, clazz)
                if (instance != null) depInstanceCache[clazz] = instance
                instance
            } catch (t: Throwable) {
                XposedHelpers.log(t)
                null
            }
        }

        @JvmStatic
        fun getViewInfo(view: View?, key: String): Any? {
            if (view == null) return null
            val info = view.getTag(viewInfoTag)
            if (info == null) return null
            @Suppress("UNCHECKED_CAST")
            val viewInfo = info as HashMap<String, Any?>
            return viewInfo[key]
        }

        @JvmStatic
        fun setViewInfo(view: View?, key: String, value: Any?) {
            if (view == null) return
            val info = view.getTag(viewInfoTag)
            val viewInfo: HashMap<String, Any?> = if (info == null) {
                val newInfo = HashMap<String, Any?>()
                view.setTag(viewInfoTag, newInfo)
                newInfo
            } else {
                @Suppress("UNCHECKED_CAST")
                info as HashMap<String, Any?>
            }
            viewInfo[key] = value
        }

        @JvmStatic
        @Throws(PendingIntent.CanceledException::class)
        fun getFreeformOptions(mContext: Context, pkgName: String, pendingIntent: PendingIntent, ignoreCheck: Boolean): Bundle? {
            if (!ignoreCheck) {
                val foregroundInfo: ForegroundInfo? = ProcessManager.getForegroundInfo()
                if (foregroundInfo != null) {
                    val topPackage = foregroundInfo.mForegroundPackageName
                    if (pkgName == topPackage) return null
                }
                val freeFormStackInfoList = MiuiFreeFormManager.getAllFreeFormStackInfosOnDisplay(
                    mContext.display?.displayId ?: 0
                )
                val freeFormCount = freeFormStackInfoList?.size ?: 0
                if (freeFormCount == 2) return null
                freeFormStackInfoList?.forEach { rootTaskInfo ->
                    if (pkgName == rootTaskInfo.packageName) return null
                }
            }
            if (!pendingIntent.isActivity) {
                val bIntent = Intent(tv.withaibuild.customiuizer.mods.GlobalActions.ACTION_PREFIX + "SetFreeFormPackage")
                bIntent.putExtra("package", pkgName)
                bIntent.setPackage("android")
                mContext.sendBroadcast(bIntent)
            }
            val options: ActivityOptions? = MiuiMultiWindowUtils.getActivityOptions(mContext, pkgName, true, false)
            if (options != null) {
                XposedHelpers.callMethod(options, "setFreeformAnimation", false)
            }
            return options?.toBundle()
        }

        @JvmStatic
        fun getFreeformIntent(pkgName: String): Intent {
            val intent = Intent()
            if (pkgName != "com.tencent.tim") {
                XposedHelpers.callMethod(intent, "addFlags", 134217728)
                XposedHelpers.callMethod(intent, "addFlags", 268435456)
                XposedHelpers.callMethod(intent, "addMiuiFlags", 256)
            }
            return intent
        }

        @JvmStatic
        fun getCPUThermalId(): Int {
            if (thermalId != -1) return thermalId
            for (i in 2 until 40 step 2) {
                try {
                    RandomAccessFile("/sys/devices/virtual/thermal/thermal_zone$i/type", "r").use { cpuReader ->
                        val sensorType = cpuReader.readLine()
                        if (sensorType != null && (sensorType.startsWith("cpu-") || sensorType.startsWith("cpu_big"))) {
                            thermalId = i
                        }
                    }
                } catch (ign: Throwable) {
                }
            }
            return thermalId
        }

        @JvmStatic
        fun replacePkgAndFrameworkValue(pkg: String, type: String, name: String, resValue: Any?) {
            if (pkg != "android") {
                MainModule.resHooks.setThemeValueReplacement("android", type, name, resValue)
            }
            MainModule.resHooks.setThemeValueReplacement(pkg, type, name, resValue)
        }

        @JvmStatic
        fun getObjectFieldByPath(target: Any?, path: String): Any? {
            if (target == null) return null
            var obj: Any? = target
            for (field in path.split('.')) {
                obj = getObjectFieldSilently(obj, field)
                if (obj == NOT_EXIST_SYMBOL) {
                    return NOT_EXIST_SYMBOL
                }
            }
            return obj
        }
    }
}
